### PR TITLE
Fixed #31211 -- Added SearchConfig expression.

### PR DESCRIPTION
**Pull request by hannseman**
https://code.djangoproject.com/ticket/31211

### CODE REVIEW

File: django/contrib/postgres/search.py [hunk 1]
```diff
@@ -33,6 +35,24 @@ def db_type(self, connection):
         return 'tsquery'
 
 
+class SearchConfig(Expression):
+    def __init__(self, config):
+        super().__init__(output_field=None)
```
Comment (charettes) [inline django/contrib/postgres/search.py]:
```suggestion
        super().__init__()
```

File: django/contrib/postgres/search.py [hunk 2]
```diff
@@ -33,6 +35,24 @@ def db_type(self, connection):
...
+        if not (config and hasattr(config, 'resolve_expression')):
+            config = Value(config)
+        self.config = config
+
+    def resolve_expression(self, query=None, allow_joins=True, reuse=None, summarize=False, for_save=False):
+        resolved = super().resolve_expression(query, allow_joins, reuse, summarize, for_save)
+        resolved.config = self.config.resolve_expression(query, allow_joins, reuse, summarize, for_save)
+        return resolved
```
Comment (charettes) [inline django/contrib/postgres/search.py]: The usual pattern is to implement `get_source_expressions` and `set_source_expressions`.

```python
def get_source_expressions(self):
    return [self.config]

def set_source_expressions(self, expressions):
    self.config, = expressions

File: django/contrib/postgres/search.py [hunk 3]
```diff
@@ -33,6 +35,24 @@ def db_type(self, connection):
         return 'tsquery'
 
 
+class SearchConfig(Expression):
+    def __init__(self, config):
+        super().__init__(output_field=None)
+        if not (config and hasattr(config, 'resolve_expression')):
```
Comment (charettes) [inline django/contrib/postgres/search.py]: The following is likely more readable given `config` works on `None` as well.

```suggestion
        if not hasattr(config, 'resolve_expression'):
```

File: django/contrib/postgres/search.py (hunk 2 above)
Comment (hannseman) [reply django/contrib/postgres/search.py]: Ah, makes sense!

**Comment by felixxm** [comment]
@hannseman Thanks :+1:

**Comment by hannseman** [comment]
@charettes @felixxm thanks for the review!

**Pull request merged**